Mathlib/RingTheory/NoetherNormalization.lean

```diff
@@ -0,0 +1,309 @@
+      leadingCoeff_pow, algebraMap_eq]
+    have : ∀ j, ((finSuccEquiv k n) ((T1 f) 1 (X j))).leadingCoeff = 1 := by
+      intro j
+      simp only [aeval_eq_bind₁, bind₁_X_right]
+      by_cases h : j = 0
+      · simp only [if_pos h, finSuccEquiv_apply, eval₂Hom_X', Fin.cases_zero, monic_X,
+          Monic.leadingCoeff]
+      · simp only [if_neg h, one_smul, map_add, map_pow]
+        obtain ⟨i, hi⟩ := Fin.exists_succ_eq.mpr h
+        rw [← hi, finSuccEquiv_X_succ, finSuccEquiv_X_zero, add_comm]
+        apply leadingCoeff_X_pow_add_C
+        simp only [add_pos_iff, ofNat_pos, true_or, pow_pos]
+    · simp only [this, one_pow, Finset.prod_const_one, mul_one]
+  · exact fun i ↦ pow_zero _
+
+/- `T` maps `f` into some polynomial of `X_0` such that the leading coeff is invertible. -/
+private lemma T_leadingcoeff (fne : f ≠ 0) :
+    ∃ c : (MvPolynomial (Fin n) k)ˣ, (c.val • (finSuccEquiv k n (T f f))).Monic := by
+  obtain ⟨v, vin, vs⟩ := Finset.exists_max_image f.support
+    (fun v ↦ (T f ((monomial v) (coeff v f))).degreeOf 0) (support_nonempty.mpr fne)
+  set h := fun w ↦ (MvPolynomial.monomial w) (coeff w f)
+  simp only [← natDegree_finSuccEquiv] at vs
+  replace vs : ∀ x ∈ f.support \ {v}, (finSuccEquiv k n ((T f) (h x))).degree <
+      (finSuccEquiv k n ((T f) (h v))).degree := by
+    intro x hx
+    obtain ⟨h1, h2⟩ := Finset.mem_sdiff.mp hx
+    apply degree_lt_degree <| lt_of_le_of_ne (vs x h1) ?_
+    simpa only [natDegree_finSuccEquiv] using T_degree f _ _ h1 vin <| ne_of_not_mem_cons h2
+  have coeff : (finSuccEquiv k n ((T f) (h v + ∑ x ∈ f.support \ {v}, h x))).leadingCoeff =
+      (finSuccEquiv k n ((T f) (h v))).leadingCoeff := by
+    simp only [map_add, map_sum]
+    rw [add_comm]
+    apply leadingCoeff_add_of_degree_lt <| (lt_of_le_of_lt <| degree_sum_le _ _) ?_
+    have h2 : h v ≠ 0 := by simpa [h] using mem_support_iff.mp vin
+    replace h2 : (finSuccEquiv k n ((T f) (h v))) ≠ 0 := fun eq ↦ h2 <|
+      by simpa only [map_eq_zero_iff _ (AlgEquiv.injective _)] using eq
+    exact (Finset.sup_lt_iff <| Ne.bot_lt (fun x ↦ h2 <| degree_eq_bot.mp x)).mpr vs
+  nth_rw 2 [← f.support_sum_monomial_coeff]
+  rw [Finset.sum_eq_add_sum_diff_singleton vin h]
+  rw [T_coeff] at coeff
+  have u : IsUnit (finSuccEquiv k n ((T f) (h v + ∑ x ∈ f.support \ {v}, h x))).leadingCoeff := by
+    simpa only [coeff, algebraMap_eq] using (mem_support_iff.mp vin).isUnit.map MvPolynomial.C
+  exact ⟨u.unit⁻¹, monic_of_isUnit_leadingCoeff_inv_smul u⟩
+
+end equivT
+
+section intmaps
+/-!
+Second we construct the following maps: Let `I` be an ideal containing `f`.
+  φ : k[X_0,...X_{n-1}] ≃ k[X_1,...X_n][X];
+  hom1 : k[X_0,...X_{n-1}] → k[X_1,...X_n][X]/φ(T(I));
+  eqv1 : k[X_1,...X_n][X]/φ(T(I)) ≃ k[X_0,...,X_n]/T(I);
+  eqv2 : k[X_0,...,X_n]/T(I) ≃ k[X_0,...,X_n]/I;
+  hom2 : k[X_0,...X_(n-1)] → k[X_0,...X_n]/I.
+hom1 is integral because φ(T(I)) contains a monic polynomial.
+hom2 is integral because it's the composition of integral maps.
```

Comment:
```suggestion
Second we construct the following maps: Let `I` be an ideal containing `f`.
  `φ : k[X_0,...X_{n-1}] ≃ k[X_1,...X_n][X]`
  `hom1 : k[X_0,...X_{n-1}] → k[X_1,...X_n][X]/φ(T(I))`
  `eqv1 : k[X_1,...X_n][X]/φ(T(I)) ≃ k[X_0,...,X_n]/T(I)`
  `eqv2 : k[X_0,...,X_n]/T(I) ≃ k[X_0,...,X_n]/I`
  `hom2 : k[X_0,...X_(n-1)] → k[X_0,...X_n]/I`
`hom1` is integral because `φ(T(I))` contains a monic polynomial.
`hom2` is integral because it's the composition of integral maps.
```
Also, can you use the actual Lean notation (instead of `≃` and such)?